brainiak/isfc.py

```diff
@@ -90,27 +143,135 @@ def isfc(D, collapse_subj=True):
         Whether to average across subjects before returning result
 
+    return_p : bool, default:False
+        Whether to use phase randomization to compute a p value for each voxel
+
+    num_perm : int, default:1000
+        Number of null samples to use for computing p values
+
+    two_sided : bool, default:False
+        Whether the p value should be one-sided (testing only for being
+        above the null) or two-sided (testing for both significantly positive
+        and significantly negative values)
+
     Returns
     -------
     ISFC : voxel by voxel ndarray
         (or voxel by voxel by subject ndarray, if collapse_subj=False)
         pearson correlation between all pairs of voxels, across subjects
+
+    p : ndarray the same shape as ISC (if return_p = True)
+        p values for each ISC value under the null distribution
     """
 
     n_vox = D.shape[0]
     n_subj = D.shape[2]
-    ISFC = np.zeros((n_vox, n_vox, n_subj))
 
-    # Loop across choice of leave-one-out subject
-    for loo_subj in range(D.shape[2]):
-        group = np.mean(D[:, :, np.arange(n_subj) != loo_subj], axis=2)
-        subj = D[:, :, loo_subj]
-        ISFC[:, :, loo_subj] = compute_correlation(group, subj)
+    if return_p:
+        n_perm = num_perm
+    else:
+        n_perm = 0
+
+    ISFC = np.zeros((n_vox, n_vox, n_subj, n_perm + 1))
+
+    for p in range(n_perm + 1):
+        # Loop across choice of leave-one-out subject
+        for loo_subj in range(D.shape[2]):
+            group = np.mean(D[:, :, np.arange(n_subj) != loo_subj], axis=2)
+            subj = D[:, :, loo_subj]
+            ISFC[:, :, loo_subj, p] = compute_correlation(group, subj)
+
+            # Symmetrize matrix
+            ISFC[:, :, loo_subj, p] = (ISFC[:, :, loo_subj, p] +
+                                       ISFC[:, :, loo_subj, p].T) / 2
 
-        # Symmetrize matrix
-        ISFC[:, :, loo_subj] = (ISFC[:, :, loo_subj] +
-                                ISFC[:, :, loo_subj].T) / 2
+        # Randomize phases of D to create next null dataset
+        D = phase_randomize(D)
 
     if collapse_subj:
         ISFC = np.mean(ISFC, axis=2)
-    return ISFC
+
+    if not return_p:
+        return np.squeeze(ISFC)
+
+    # Compute maximum/minimum ISFC in each null dataset
+    if collapse_subj:
+        max_null = np.max(ISFC[:, :, 1:], axis=(0, 1))
+        min_null = np.min(ISFC[:, :, 1:], axis=(0, 1))
+        ISFC = ISFC[:, :, 0]
+    else:
+        max_null = np.max(ISFC[:, :, :, 1:], axis=(0, 1, 2))
+        min_null = np.min(ISFC[:, :, 1:], axis=(0, 1))
+        ISFC = ISFC[:, :, :, 0]
+
+    # Compute where the true values fall on the null distribution
+    max_null_ecdf = ecdf(max_null)
+    if two_sided:
+        min_null_ecdf = ecdf(min_null)
+        p = 2 * np.minimum(1 - max_null_ecdf(ISFC), min_null_ecdf(ISFC))
+        p = np.minimum(p, 1)
+    else:
+        p = 1 - max_null_ecdf(ISFC)
+    return ISFC, p
+
+
+def phase_randomize(D):

```

Comment:
This function is not really ISFC specific. I think it fits better in `utils/utils.py`.